File: src/main/java/com/snowflake/snowpark_java/types/Geography.java
package com.snowflake.snowpark_java.types;

import java.io.IOException;
import java.io.Serializable;
import java.io.UncheckedIOException;

/**
 * Java representation of Snowflake Geography data.
 *
 * @since 0.8.0
 */
public class Geography implements Serializable {

  private final String data;

  private Geography(String data) {
    if (data == null) {
      throw new UncheckedIOException(
          new IOException("Cannot create geography object from null input"));
    }
    this.data = data;
  }

  /**
   * Checks whether two Geography object are equal.
   *
   * @param other A Geography object
   * @return true if these two object are equal
   */
  @Override
  public boolean equals(Object other) {
    if (other instanceof Geography) {
      return data.equals(((Geography) other).data);
    }
    return false;
  }

  /**
   * Calculates the hash code of this Geography Object.
   *
   * @return An int number representing the hash code value
   * @since 0.9.0
   */
  @Override
  public int hashCode() {
    return this.data.hashCode();
  }

  /**
   * Converts this Geography object to a GeoJSON string.
   *
   * @return A GeoJSON string.
   * @since 0.8.0
   */
  public String asGeoJSON() {
    return data;
  }

  /**
   * Converts this Geography object to a String value. alias of {@code asGeoJson} function.
   *
   * @return A String value.
   */
  @Override
  public String toString() {
    return asGeoJSON();
  }

  /**
   * Creates a Geography object from a GeoJSON string.
   *
   * @param g GeoJSON String
   * @return a new Geography object
   * @since 0.8.0
   */
  public static Geography fromGeoJSON(String g) {
    return new Geography(g);
  }
}

File: v2/src/main/scala/com/snowflake/snowpark/types/Geography.scala
package com.snowflake.snowpark.types

import java.io.{IOException, UncheckedIOException}

/**
 * Companion object of Geography class.
 */
object Geography {

  /**
   * Creates a Geography class from a GeoJSON string
   *
   * @param g
   *   GeoJSON string
   * @return
   *   a Geography class
   * @since 0.2.0
   */
  def fromGeoJSON(g: String): Geography = new Geography(g)
}

/**
 * Scala representation of Snowflake Geography data. Only support GeoJSON format.
 *
 * @since 0.2.0
 */
class Geography private (private val stringData: String) {
  if (stringData == null) throwNullInputError()

  /**
   * Returns whether the Geography object equals to the input object.
   *
   * @return
   *   GeoJSON string
   * @since 0.2.0
   */
  override def equals(obj: Any): Boolean = {
    obj match {
      case g: Geography => stringData.equals(g.stringData)
      case _ => false
    }
  }

  /**
   * Returns the hashCode of the stored GeoJSON string.
   *
   * @return
   *   hash code
   * @since 0.2.0
   */
  override def hashCode(): Int = stringData.hashCode

  private def throwNullInputError(): Unit =
    throw new UncheckedIOException(
      new IOException("Cannot create geography object from null input"))

  /**
   * Returns the underling string data for GeoJSON.
   *
   * @return
   *   GeoJSON string
   * @since 0.2.0
   */
  def asGeoJSON(): String = stringData

  /**
   * Returns the underling string data for GeoJSON.
   *
   * @return
   *   GeoJSON string
   * @since 0.2.0
   */
  def getString: String = stringData

  /**
   * Returns the underling string data for GeoJSON.
   *
   * @return
   *   GeoJSON string
   * @since 0.2.0
   */
  override def toString: String = stringData
}

/**
 * Geography data type. This maps to GEOGRAPHY data type in Snowflake.
 * @since 0.2.0
 */
object GeographyType extends DataType {
  override def toString: String = {
    s"GeographyType"
  }

  override private[snowpark] def schemaString: String = s"Geography"
}
